src/IO/ReadBufferFromS3.cpp

```diff
@@ -84,44 +78,49 @@ bool ReadBufferFromS3::nextImpl()
     ProfileEvents::increment(ProfileEvents::S3ReadMicroseconds, watch.elapsedMicroseconds());
     if (!next_result)
         return false;
-    internal_buffer = impl->buffer();
+
+    working_buffer = internal_buffer = impl->buffer();
+    pos = working_buffer.begin();
 
     ProfileEvents::increment(ProfileEvents::S3ReadBytes, internal_buffer.size());
 
-    working_buffer = internal_buffer;
+    offset += working_buffer.size();
+
     return true;
 }
 
 off_t ReadBufferFromS3::seek(off_t offset_, int whence)
 {
-    if (impl)
-        throw Exception("Seek is allowed only before first read attempt from the buffer.", ErrorCodes::CANNOT_SEEK_THROUGH_FILE);
-
     if (whence != SEEK_SET)
         throw Exception("Only SEEK_SET mode is allowed.", ErrorCodes::CANNOT_SEEK_THROUGH_FILE);
 
     if (offset_ < 0)
         throw Exception("Seek position is out of bounds. Offset: " + std::to_string(offset_), ErrorCodes::SEEK_POSITION_OUT_OF_BOUND);
 
+    if (impl)
+    {
+        impl.reset();
+        pos = working_buffer.end();
+    }
+
     offset = offset_;
 
     return offset;
 }
 
 off_t ReadBufferFromS3::getPosition()
 {
-    return offset + count();
+    return offset - available();
```

Comment:
Why `bytes`/`count()` is unreliable in case when there is no seeks, so we have to count it ourselves? Can you make a simple test on that? I was not able to reproduce it on S3 table function.

Reply:
It reproduces when we use DiskS3 in `SeekAvoidingReadBuffer` -> `ReadIndirectBufferFromRemoteFS` -> `ReadBufferFromS3` chain.
In this case we shift position and count bytes on the top buffer of the chain `SeekAvoidingReadBuffer` after corresponding `swap(*buffer)` calls.
I was able to reproduce it with `test_s3_disk_restart_during_load`.